File: src/web/gauges.ts
/**
 * Usage columns for the resource tables.
 *
 * A pod list that shows READY and AGE but not what the thing is *doing* sends
 * you to `kubectl top` in another window, so pods, the workloads that own them
 * and nodes all get live CPU/memory columns with a bar.
 *
 * The bar's denominator is the honest part. A container with a limit is drawn
 * against its limit and marked at its request — the two numbers that decide
 * whether it gets throttled or OOM-killed. A container with no limit has no
 * true ceiling, so the column falls back to the largest value in this table:
 * the bar then means "relative to the busiest row here", which is a comparison,
 * not a promise. Both are honest; a bar with an invented ceiling would not be.
 */

import { type K8sObject, listResources, topNodes, topPods } from "../kubectl.ts";
import { workloadSelector } from "../format.ts";
import { formatBytes, formatCpu, parseQuantity, percent } from "../quantity.ts";
import { type Gauge, addGauge, emptyGauge, podAllocation, selectorMatches } from "../usage.ts";

export interface Meter {
    pct: number | null;
    mark: number | null;
}

export interface UsageColumns {
    /** Column labels to splice into the kind's own columns. */
    columns: string[];
    /** Column label to insert before; falls back to the end of the row. */
    insertBefore: string;
    /** Cells and bars per object, keyed "ns/name". */
    byKey: Map<string, { cells: string[]; meters: (Meter | null)[] }>;
}

const key = (obj: K8sObject) => `${obj.metadata?.namespace ?? ""}/${obj.metadata?.name ?? ""}`;

/** Kinds whose rows carry usage bars. */
export const USAGE_KINDS = new Set([
    "pods",
    "deployments",
    "statefulsets",
    "daemonsets",
    "replicasets",
    "jobs",
    "nodes",
]);

function meterFor(gauge: Gauge, fallbackMax: number): Meter {
    const ceiling = gauge.limits > 0 ? gauge.limits : fallbackMax;
    return {
        pct: gauge.used === null ? null : percent(gauge.used, ceiling),
        mark: gauge.requests > 0 ? percent(gauge.requests, ceiling) : null,
    };
}

/** Largest value a column will have to draw, so bars share one scale. */
function ceilingOf(gauges: Gauge[], pick: (g: Gauge) => number): number {
    let max = 0;
    for (const g of gauges) {
        max = Math.max(max, pick(g), g.used ?? 0);
    }
    return max > 0 ? max * 1.15 : 1;
}

export async function usageColumns(
    kindName: string,
    items: K8sObject[],
    context: string,
    namespace: string | null,
): Promise<UsageColumns | null> {
    if (!USAGE_KINDS.has(kindName) || items.length === 0) return null;
    try {
        if (kindName === "nodes") return await nodeUsage(items, context);
        return await podsideUsage(kindName, items, context, namespace);
    } catch {
        // Metrics are a bonus column, never a reason a table fails to load.
        return null;
    }
}

/** Pods, and anything whose usage is the sum of the pods it owns. */
async function podsideUsage(
    kindName: string,
    items: K8sObject[],
    context: string,
    namespace: string | null,
): Promise<UsageColumns | null> {
    const ns = namespace ?? undefined;
    const [top, pods] = await Promise.all([
        topPods(context, ns).catch(() => new Map()),
        kindName === "pods"
            ? Promise.resolve(items)
            : listResources("pods", { context, namespace: ns }).catch(() => [] as K8sObject[]),
    ]);
    const sample = (pod: K8sObject): { cpu: number; mem: number } | null => {
        const m =
            top.get(`${pod.metadata?.namespace ?? ""}/${pod.metadata?.name ?? ""}`) ?? top.get(pod.metadata?.name ?? "");
        return m ? { cpu: parseQuantity(m.cpu), mem: parseQuantity(m.memory) } : null;
    };

    const gauges = new Map<string, { cpu: Gauge; mem: Gauge }>();
    for (const obj of items) {
        if (kindName === "pods") {
            const alloc = podAllocation(obj);
            const live = sample(obj);
            gauges.set(key(obj), {
                cpu: { ...alloc.cpu, used: live ? live.cpu : null },
                mem: { ...alloc.mem, used: live ? live.mem : null },
            });
            continue;
        }
        // A workload's pods are the ones its own selector matches, in its own
        // namespace. Walking ownerReferences would need the intermediate
        // ReplicaSets — one more list call for the same answer.
        const selector = (obj.spec as { selector?: unknown })?.selector;
        let cpu = emptyGauge();
        let mem = emptyGauge();
        for (const pod of pods) {
            if (pod.metadata?.namespace !== obj.metadata?.namespace) continue;
            if (!selectorMatches(pod, selector)) continue;
            const alloc = podAllocation(pod);
            const live = sample(pod);
            cpu = addGauge(cpu, { ...alloc.cpu, used: live ? live.cpu : null });
            mem = addGauge(mem, { ...alloc.mem, used: live ? live.mem : null });
        }
        gauges.set(key(obj), { cpu, mem });
    }

    const all = [...gauges.values()];
    const cpuCeiling = ceilingOf(all.map((g) => g.cpu), (g) => g.limits);
    const memCeiling = ceilingOf(all.map((g) => g.mem), (g) => g.limits);

    const byKey = new Map<string, { cells: string[]; meters: (Meter | null)[] }>();
    for (const [k, g] of gauges) {
        byKey.set(k, {
            cells: [
                g.cpu.used === null ? "—" : formatCpu(g.cpu.used),
                g.mem.used === null ? "—" : formatBytes(g.mem.used),
            ],
            meters: [meterFor(g.cpu, cpuCeiling), meterFor(g.mem, memCeiling)],
        });
    }
    return { columns: ["CPU USAGE", "MEMORY USAGE"], insertBefore: "AGE", byKey };
}

/** Nodes: usage against capacity, and how much of it is already promised. */
async function nodeUsage(items: K8sObject[], context: string): Promise<UsageColumns> {
    const [top, pods] = await Promise.all([
        topNodes(context).catch(() => new Map()),
        listResources("pods", { context }).catch(() => [] as K8sObject[]),
    ]);

    const requested = new Map<string, { cpu: number; mem: number; pods: number }>();
    for (const pod of pods) {
        const node = (pod.spec as { nodeName?: string })?.nodeName;
        const phase = (pod.status as { phase?: string })?.phase;
        if (!node || phase === "Succeeded" || phase === "Failed") continue;
        const alloc = podAllocation(pod);
        const prev = requested.get(node) ?? { cpu: 0, mem: 0, pods: 0 };
        requested.set(node, {
            cpu: prev.cpu + alloc.cpu.requests,
            mem: prev.mem + alloc.mem.requests,
            pods: prev.pods + 1,
        });
    }

    const byKey = new Map<string, { cells: string[]; meters: (Meter | null)[] }>();
    for (const node of items) {
        const name = node.metadata?.name ?? "";
        const status = node.status as { allocatable?: Record<string, string>; capacity?: Record<string, string> };
        const alloc = status?.allocatable ?? status?.capacity ?? {};
        const cpuCap = parseQuantity(alloc.cpu);
        const memCap = parseQuantity(alloc.memory);
        const podCap = parseQuantity(alloc.pods);
        const live = top.get(name);
        const req = requested.get(name) ?? { cpu: 0, mem: 0, pods: 0 };
        const cpuUsed = live ? parseQuantity(live.cpu) : null;
        const memUsed = live ? parseQuantity(live.memory) : null;
        const cpuAlloc = percent(req.cpu, cpuCap);
        const memAlloc = percent(req.mem, memCap);

        byKey.set(key(node), {
            cells: [
                podCap > 0 ? `${req.pods}/${podCap}` : String(req.pods),
                cpuUsed === null ? "—" : formatCpu(cpuUsed),
                cpuAlloc === null ? "—" : `${Math.round(cpuAlloc)}%`,
                memUsed === null ? "—" : formatBytes(memUsed),
                memAlloc === null ? "—" : `${Math.round(memAlloc)}%`,
            ],
            meters: [
                { pct: percent(req.pods, podCap), mark: null },
                { pct: cpuUsed === null ? null : percent(cpuUsed, cpuCap), mark: cpuAlloc },
                { pct: cpuAlloc, mark: null },
                { pct: memUsed === null ? null : percent(memUsed, memCap), mark: memAlloc },
                { pct: memAlloc, mark: null },
            ],
        });
    }

    return {
        columns: ["PODS", "CPU USAGE", "CPU ALLOC", "MEMORY USAGE", "MEM ALLOC"],
        insertBefore: "ROLES",
        byKey,
    };
}

File: src/internal/server/gauges.go
package server

import (
	"fmt"
	"math"

	"github.com/notshekhar/digg/src/internal/kube"
	"github.com/notshekhar/digg/src/internal/model"
)

// Usage columns for the resource tables. Port of src/web/gauges.ts.
//
// A pod list that shows READY and AGE but not what the thing is *doing* sends
// you to `kubectl top` in another window, so pods, the workloads that own them
// and nodes all get live CPU/memory columns with a bar.
//
// The bar's denominator is the honest part. A container with a limit is drawn
// against its limit and marked at its request — the two numbers that decide
// whether it gets throttled or OOM-killed. A container with no limit has no
// true ceiling, so the column falls back to the largest value in this table:
// the bar then means "relative to the busiest row here", which is a comparison,
// not a promise. Both are honest; a bar with an invented ceiling would not be.

// Meter is one bar: a fill and an optional tick.
type Meter struct {
	Pct  *float64 `json:"pct"`
	Mark *float64 `json:"mark"`
}

type usageCell struct {
	Cells  []string
	Meters []*Meter
}

// UsageColumns are the extra columns spliced into a kind's own.
type UsageColumns struct {
	// Columns are the labels to splice in.
	Columns []string
	// InsertBefore is the column label to insert before; falls back to the end
	// of the row.
	InsertBefore string
	// ByKey holds cells and bars per object, keyed "ns/name".
	ByKey map[string]usageCell
}

// UsageKinds are the kinds whose rows carry usage bars.
var UsageKinds = map[string]bool{
	"pods":         true,
	"deployments":  true,
	"statefulsets": true,
	"daemonsets":   true,
	"replicasets":  true,
	"jobs":         true,
	"nodes":        true,
}

func objKey(o *model.Obj) string {
	return o.GetNamespace() + "/" + o.GetName()
}

func pctPtr(used, total float64) *float64 {
	if v, ok := model.Percent(used, total); ok {
		return &v
	}
	return nil
}

func meterFor(g model.Gauge, fallbackMax float64) *Meter {
	ceiling := g.Limits
	if ceiling <= 0 {
		ceiling = fallbackMax
	}
	m := &Meter{}
	if g.Used != nil {
		m.Pct = pctPtr(*g.Used, ceiling)
	}
	if g.Requests > 0 {
		m.Mark = pctPtr(g.Requests, ceiling)
	}
	return m
}

// ceilingOf is the largest value a column will have to draw, so bars share one
// scale.
func ceilingOf(gauges []model.Gauge) float64 {
	max := 0.0
	for _, g := range gauges {
		max = math.Max(max, g.Limits)
		if g.Used != nil {
			max = math.Max(max, *g.Used)
		}
	}
	if max > 0 {
		return max * 1.15
	}
	return 1
}

// usageColumnsFor builds the usage columns for a listing, or nil when the kind
// has none. Metrics are a bonus column, never a reason a table fails to load,
// so every failure below returns nil rather than an error.
func usageColumnsFor(cl *kube.Cluster, kindName string, items []model.Obj, namespace string) *UsageColumns {
	if !UsageKinds[kindName] || len(items) == 0 {
		return nil
	}
	if kindName == "nodes" {
		return nodeUsage(cl, items)
	}
	return podsideUsage(cl, kindName, items, namespace)
}

// podsideUsage covers pods, and anything whose usage is the sum of the pods it
// owns.
func podsideUsage(cl *kube.Cluster, kindName string, items []model.Obj, namespace string) *UsageColumns {
	top := cl.TopPods(namespace, "")

	var pods []model.Obj
	if kindName == "pods" {
		pods = items
	} else {
		var err error
		pods, err = cl.List("pods", kube.ListOptions{Namespace: namespace})
		if err != nil {
			pods = nil
		}
	}

	sample := func(pod *model.Obj) (float64, float64, bool) {
		m, ok := top[pod.GetNamespace()+"/"+pod.GetName()]
		if !ok {
			m, ok = top[pod.GetName()]
		}
		if !ok {
			return 0, 0, false
		}
		return model.ParseQuantity(m.CPU), model.ParseQuantity(m.Memory), true
	}

	type pair struct{ cpu, mem model.Gauge }
	gauges := map[string]pair{}
	order := make([]string, 0, len(items))

	for i := range items {
		o := &items[i]
		k := objKey(o)
		order = append(order, k)

		if kindName == "pods" {
			alloc := model.PodAllocation(o)
			cpu, mem := alloc.CPU, alloc.Mem
			if c, m, ok := sample(o); ok {
				cpu.Used, mem.Used = &c, &m
			}
			gauges[k] = pair{cpu, mem}
			continue
		}

		// A workload's pods are the ones its own selector matches, in its own
		// namespace. Walking ownerReferences would need the intermediate
		// ReplicaSets — one more list call for the same answer.
		selector := nestedMap(o, "spec", "selector")
		var cpu, mem model.Gauge
		for j := range pods {
			p := &pods[j]
			if p.GetNamespace() != o.GetNamespace() {
				continue
			}
			if !model.SelectorMatches(p, selector) {
				continue
			}
			alloc := model.PodAllocation(p)
			pc, pm := alloc.CPU, alloc.Mem
			if c, m, ok := sample(p); ok {
				pc.Used, pm.Used = &c, &m
			}
			cpu = model.AddGauge(cpu, pc)
			mem = model.AddGauge(mem, pm)
		}
		gauges[k] = pair{cpu, mem}
	}

	cpuAll := make([]model.Gauge, 0, len(gauges))
	memAll := make([]model.Gauge, 0, len(gauges))
	for _, g := range gauges {
		cpuAll = append(cpuAll, g.cpu)
		memAll = append(memAll, g.mem)
	}
	cpuCeiling, memCeiling := ceilingOf(cpuAll), ceilingOf(memAll)

	byKey := map[string]usageCell{}
	for _, k := range order {
		g := gauges[k]
		cpuText, memText := "—", "—"
		if g.cpu.Used != nil {
			cpuText = model.FormatCPU(*g.cpu.Used)
		}
		if g.mem.Used != nil {
			memText = model.FormatBytes(*g.mem.Used)
		}
		byKey[k] = usageCell{
			Cells:  []string{cpuText, memText},
			Meters: []*Meter{meterFor(g.cpu, cpuCeiling), meterFor(g.mem, memCeiling)},
		}
	}
	return &UsageColumns{
		Columns:      []string{"CPU USAGE", "MEMORY USAGE"},
		InsertBefore: "AGE",
		ByKey:        byKey,
	}
}

// nodeUsage covers nodes: usage against capacity, and how much of it is already
// promised.
func nodeUsage(cl *kube.Cluster, items []model.Obj) *UsageColumns {
	top := cl.TopNodes()
	pods, err := cl.List("pods", kube.ListOptions{})
	if err != nil {
		pods = nil
	}

	type req struct {
		cpu, mem float64
		pods     int
	}
	requested := map[string]req{}
	for i := range pods {
		p := &pods[i]
		node := nestedString(p, "spec", "nodeName")
		phase := nestedString(p, "status", "phase")
		if node == "" || phase == "Succeeded" || phase == "Failed" {
			continue
		}
		alloc := model.PodAllocation(p)
		prev := requested[node]
		requested[node] = req{
			cpu:  prev.cpu + alloc.CPU.Requests,
			mem:  prev.mem + alloc.Mem.Requests,
			pods: prev.pods + 1,
		}
	}

	byKey := map[string]usageCell{}
	for i := range items {
		node := &items[i]
		name := node.GetName()

		alloc := nestedStringMap(node, "status", "allocatable")
		if len(alloc) == 0 {
			alloc = nestedStringMap(node, "status", "capacity")
		}
		cpuCap := model.ParseQuantity(alloc["cpu"])
		memCap := model.ParseQuantity(alloc["memory"])
		podCap := model.ParseQuantity(alloc["pods"])

		r := requested[name]
		cpuAlloc := pctPtr(r.cpu, cpuCap)
		memAlloc := pctPtr(r.mem, memCap)

		var cpuUsed, memUsed *float64
		if live, ok := top[name]; ok {
			c := model.ParseQuantity(live.CPU)
			m := model.ParseQuantity(live.Memory)
			cpuUsed, memUsed = &c, &m
		}

		podCell := fmt.Sprintf("%d", r.pods)
		if podCap > 0 {
			podCell = fmt.Sprintf("%d/%s", r.pods, model.FormatCPU(podCap))
		}
		cpuCell, memCell := "—", "—"
		if cpuUsed != nil {
			cpuCell = model.FormatCPU(*cpuUsed)
		}
		if memUsed != nil {
			memCell = model.FormatBytes(*memUsed)
		}

		cells := []string{podCell, cpuCell, pctCell(cpuAlloc), memCell, pctCell(memAlloc)}

		meters := []*Meter{
			{Pct: pctPtr(float64(r.pods), podCap)},
			{Mark: cpuAlloc},
			{Pct: cpuAlloc},
			{Mark: memAlloc},
			{Pct: memAlloc},
		}
		if cpuUsed != nil {
			meters[1].Pct = pctPtr(*cpuUsed, cpuCap)
		}
		if memUsed != nil {
			meters[3].Pct = pctPtr(*memUsed, memCap)
		}

		byKey[objKey(node)] = usageCell{Cells: cells, Meters: meters}
	}

	return &UsageColumns{
		Columns:      []string{"PODS", "CPU USAGE", "CPU ALLOC", "MEMORY USAGE", "MEM ALLOC"},
		InsertBefore: "ROLES",
		ByKey:        byKey,
	}
}

func pctCell(p *float64) string {
	if p == nil {
		return "—"
	}
	return fmt.Sprintf("%d%%", int64(math.Round(*p)))
}
